### PR TITLE
Download files locally using `response.iter_content`

### LINKED ISSUE #4157: Review how to write to a file from generator
Logging one of the graphite suggestions that does make sense to discuss. The change suggested is to change `Path.write_bytes` with something like
```python
with open(archive_path, 'wb') as f:
    for chunk in response.iter_content(chunk_size=8192):
        f.write(chunk)
```
for itterables in this case `requests.Response`. Apart from the magic number it does make sense to use something along those lines. Anyone have references for other variants of this? I encountered a few of them when I looked at this, will try to track those back.

### LINKED ISSUE #4214: Archive in `discover` raises `Unknown archive format`
tmt version:
tmt-1.59.0-2.fc42.noarch

discover phase:
 ```
 discover:
    - name: brew-build-install
      how: fmf
      url: https://pkgs.devel.redhat.com/git/tests/distribution/snapshot/distribution-master.tar.gz
      url-content-type: archive
      test:
        - /distribution/brew-build-install

    - name: ovs_bond_function
      how: fmf
      url: https://gitlab.cee.redhat.com/mhou/kernel/-/archive/master/kernel-master.tar.gz
      url-content-type: archive
      test:
        - /kernel/networking/rt-kernel/ovs_bond_function
      ref: master
```

command output
```
tmt --context distro=RHEL-10.0-updates-20251019.0  run -e @./cx6dx_dell750_32.txt -dv
Using tree '/home/houminxi/code/fdp_tmt'.
/var/tmp/tmt/run-001
tmt version: 1.59.0
tmt command line: /usr/bin/tmt --context distro=RHEL-10.0-updates-20251019.0 run -e @./cx6dx_dell750_32.txt -dv
User does not have sudo access, we assume everything is pre-setup.
Run data not found.
Enabled steps: report, prepare, execute, cleanup, finish, provision and discover
Workdir '/var/tmp/tmt/run-001/ovs_bond_function/plan' created.
Found 1 plan.

/ovs_bond_function/plan
summary: ovs bond function test
    Workdir '/var/tmp/tmt/run-001/ovs_bond_function/plan/discover' created.
    Workdir '/var/tmp/tmt/run-001/ovs_bond_function/plan/provision' created.
    Workdir '/var/tmp/tmt/run-001/ovs_bond_function/plan/prepare' created.
    Workdir '/var/tmp/tmt/run-001/ovs_bond_function/plan/execute' created.
    Workdir '/var/tmp/tmt/run-001/ovs_bond_function/plan/report' created.
    Report step always force mode enabled.
    Workdir '/var/tmp/tmt/run-001/ovs_bond_function/plan/report' created.
    Workdir '/var/tmp/tmt/run-001/ovs_bond_function/plan/finish' created.
    Workdir '/var/tmp/tmt/run-001/ovs_bond_function/plan/cleanup' created.
    discover
    workdir: /var/tmp/tmt/run-001/ovs_bond_function/plan/discover
        how: fmf
        name: brew-build-install
        Workdir '/var/tmp/tmt/run-001/ovs_bond_function/plan/discover/brew-build-install' created.
        url: https://pkgs.devel.redhat.com/git/tests/distribution/snapshot/distribution-master.tar.gz
        Downloading 'https://pkgs.devel.redhat.com/git/tests/distribution/snapshot/distribution-master.tar.gz' and extracting to '/var/tmp/tmt/run-001/ovs_bond_function/plan/discover/brew-build-install/tests'.
    Suspending step 'discover'.
    Suspending step 'provision'.
    Suspending step 'prepare'.
    Suspending step 'execute'.
    Suspending step 'finish'.
    report
    workdir: /var/tmp/tmt/run-001/ovs_bond_function/plan/report
        how: html
        Workdir '/var/tmp/tmt/run-001/ovs_bond_function/plan/report/default-0' created.
        output: /var/tmp/tmt/run-001/ovs_bond_function/plan/report/default-0/index.html
        summary: no results found
    cleanup
    workdir: /var/tmp/tmt/run-001/ovs_bond_function/plan/cleanup
        warn: Nothing to cleanup, no guests provisioned.

plan failed

The exception was caused by 1 earlier exceptions

Cause number 1:

    Unknown archive format '/var/tmp/tmt/run-001/ovs_bond_function/plan/discover/brew-build-install/"distribution-master.tar.gz"'

```

Discover step.yaml

```
cat step.yaml 
status: todo
data:
  - name: brew-build-install
    how: fmf
    order: 50
    when: []
    summary:
    where: []
    dist-git-source: false
    dist-git-type:
    dist-git-download-only: false
    dist-git-install-builddeps: false
    dist-git-require: []
    require-test: []
    url: 
        https://pkgs.devel.redhat.com/git/tests/distribution/snapshot/distribution-master.tar.gz
    url-content-type: archive
    ref:
    path:
    test:
      - /distribution/brew-build-install
    link: []
    filter: []
    include: []
    exclude: []
    modified-only: false
    modified-url:
    modified-ref:
    dist-git-init: false
    dist-git-remove-fmf-root: false
    dist-git-merge: false
    dist-git-extract:
    sync-repo: false
    fmf-id: false
    prune: false
    adjust-tests: []
    upgrade-path:
    repository:
    revision:
    __class__:
        module: tmt.steps.discover.fmf
        name: DiscoverFmfStepData
  - name: ovs_bond_function
    how: fmf
    order: 50
    when: []
    summary:
    where: []
    dist-git-source: false
    dist-git-type:
    dist-git-download-only: false
    dist-git-install-builddeps: false
    dist-git-require: []
    require-test: []
    url: 
        https://gitlab.cee.redhat.com/mhou/kernel/-/archive/master/kernel-master.tar.gz
    url-content-type: archive
    ref: master
    path:
    test:
      - /kernel/networking/rt-kernel/ovs_bond_function
    link: []
    filter: []
    include: []
    exclude: []
    modified-only: false
    modified-url:
    modified-ref:
    dist-git-init: false
    dist-git-remove-fmf-root: false
    dist-git-merge: false
    dist-git-extract:
    sync-repo: false
    fmf-id: false
    prune: false
    adjust-tests: []
    upgrade-path:
    repository:
    revision:
    __class__:
        module: tmt.steps.discover.fmf
        name: DiscoverFmfStepData

```

### PR DESCRIPTION
**Pull request by LecrisUT**
Purpose of this change is to iteratively download and write the file content to avoid clogging the memory.

Pull Request Checklist

* [x] implement the feature

Closes #4157
Closes #4215
Closes #4214

### CODE REVIEW

File: tmt/utils/url.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        response = session.get(url, stream=True)
+    response.raise_for_status()
+    if destination.is_dir():
+        if "Content-Disposition" in response.headers:
+            match = re.findall("filename=(.+)", response.headers["Content-Disposition"])
+            if not match:
+                raise tmt.utils.GeneralError("Could not determine filename from header")
+            file_name = match[0]
+        else:
+            file_name = response.url.split("/")[-1]
```
Comment (LecrisUT) [inline tmt/utils/url.py:29]: One thing we can do here is to use `urllib` which might handled `Content-Disposition` natively, needs researching

Comment (LecrisUT) [reply tmt/utils/url.py:29]: Nope `urllib3` did not handle it any more nicely. `email.message` seemed to have helped though.

File: tmt/utils/url.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    else:
+        if destination.exists():
+            raise tmt.utils.GeneralError(f"Destination file already exists: {destination}")
+        file_name = destination.name
+        destination = destination.parent
+    assert isinstance(file_name, str)  # Narrow type
+    download_file = destination / file_name
+    with download_file.open(mode="wb") as f:
+        for chunk in response.iter_content(chunk_size=None):
```
[the remark this answers is not in the document]
Comment (LecrisUT) [reply tmt/utils/url.py:38]: Absolutely false!!!
```
        chunk_size must be of type int or None. A value of None will
        function differently depending on the value of `stream`.
        stream=True will read data as it arrives in whatever size the
        chunks are received. If stream=False, data is returned as
        a single chunk.
```

File: tmt/utils/url.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+def download(url: str, destination: Path, *, logger: tmt.log.Logger) -> Path:
+    logger.debug(f"Downloading '{url}'.")
+    with tmt.utils.retry_session() as session:
+        response = session.get(url, stream=True)
+    response.raise_for_status()
+    if destination.is_dir():
+        file_name = None
+        if "Content-Disposition" in response.headers:
+            # We use email.message to parse the Content-Disposition header since the regex can get
+            # quite complicate. We do not support `filename*` with this yet, but that is even more
```
Comment (skycastlelily) [inline tmt/utils/url.py:21]: hmmm, just want to make sure have you considered use requests-toolbelt's ContentDispositionHeader? I ask claude whether it would be better,and here is its response:

● Yes, requests-toolbelt's ContentDispositionHeader would be a better solution. Here's why:

  Advantages of requests-toolbelt

  1. Purpose-built: Specifically designed for HTTP Content-Disposition headers
  2. RFC 6266 compliant: Follows the actual HTTP standard for Content-Disposition
  3. Better edge case handling: Handles encoded filenames, multiple parameters, etc.
  4. Common in ecosystem: Widely used in HTTP-related Python projects

  Usage Example

  from requests_toolbelt import ContentDispositionHeader

  header_value = response.headers.get('Content-Disposition')
  if header_value:
      cd_header = ContentDispositionHeader.parse(header_value)
      filename = cd_header.filename

  Current Code Limitations

  The email.message approach has gaps:
  - No support for filename* (RFC 5987 encoded filenames)
  - May not handle HTTP-specific quirks correctly
  - Could break on edge cases where email vs HTTP parsing differs

  But There's a Catch

  requests-toolbelt would be an additional dependency. Looking at the current code, the author likely chose email.message because:
  - It's in Python's standard library (no new dependency)
  - Works for 90% of common cases
  - Better than the previous regex approach

  Trade-off Analysis

  - Current approach: Good enough, no dependencies, some edge case issues
  - requests-toolbelt: More correct, additional dependency, better long-term

Comment (LecrisUT) [reply tmt/utils/url.py:21]: Afaict, requests-toolbelt does not support `filename*` [^1], nor am I confident about that regex pattern. It may be an option if we properly stress-test it and get it to support that format.

---

[^1]: https://github.com/requests/toolbelt/issues/379

> * May not handle HTTP-specific quirks correctly
> 
> * Could break on edge cases where email vs HTTP parsing differs

Does not apply, we simply use it for parsing that specific format in the header.

> * No support for filename* (RFC 5987 encoded filenames)

Yes and it is commented about that. We still don't know which library will support it, or which one is already complete and just needs modernization. A broader investigation is needed.

> requests-toolbelt would be an additional dependency.

Not an issue, we just need to make sure it is packaged and up to date in epel9, which this one partially is.

> from requests_toolbelt import ContentDispositionHeader
> 
> header_value = response.headers.get('Content-Disposition') if header_value: cd_header = ContentDispositionHeader.parse(header_value) filename = cd_header.filename

Hallucination (big time) [^2]. If it can gather references of what libraries do have partial support for Content-Disposition, it might be helpful to dig into those. I so far only checked a few:
- `content-disposition`: does the opposite (set instead of get)
- `rfc6266-*` severely outdated. An ok approach, but needs modernization
- stackexchange suggested regex: as we just saw it is way more complicated then we first thought so let's not carry that burden ourselves

[^2]: https://github.com/requests/toolbelt/blob/bcd5f7be229e14089052be7e3b527ebcea0ae7b8/requests_toolbelt/downloadutils/stream.py#L27

Comment (skycastlelily) [reply tmt/utils/url.py:21]: yeah, according to [_get_filename](https://github.com/requests/toolbelt/blob/bcd5f7be229e14089052be7e3b527ebcea0ae7b8/requests_toolbelt/downloadutils/stream.py#L18), requests-toolbelt does not support filename*, sorry for put the misleading comment here. 

> rfc6266-* severely outdated

yeah, ai also recommend [rfc6266-content-disposition.parse_headers](https://github.com/g2p/rfc6266/blob/master/rfc6266.py#L175) for filename*, which seems true, but according to the[ github page activities](https://github.com/g2p/rfc6266), it's definitely outdated.

Then I asked further, it offers werkzeug.http.parse_options_header

```
-import email.message

+from werkzeug.http import parse_options_header
 
 
 def download(url: str, destination: Path, *, logger: tmt.log.Logger) -> Path:
@@ -17,14 +16,10 @@ def download(url: str, destination: Path, *, logger: tmt.log.Logger) -> Path:
     if destination.is_dir():
         file_name = None
         if "Content-Disposition" in response.headers:
-            # We use email.message to parse the Content-Disposition header since the regex can get
-            # quite complicate. We do not support `filename*` with this yet, but that is even more
-            # complicated to handle. Waiting on a better library to handle it.
-            # See: https://stackoverflow.com/a/12017573
-            _msg_handler = email.message.Message()
-            _msg_handler["Content-Disposition"] = response.headers["Content-Disposition"]
-            # If this one fails, it should return a None object, then try to get from url instead
-            file_name = _msg_handler.get_filename()
+            # Use werkzeug to parse the Content-Disposition header, which supports filename*
+            _, options = parse_options_header(response.headers["Content-Disposition"])
+            # Try filename* first (RFC 6266), then fallback to filename
+            file_name = options.get("filename*") or options.get("filename")

```

I asked it extend the test to cover filename* senario, and here is the[ lots of  testcases](https://lnie.fedorapeople.org/test_url.py) it extended,  and all are passed.

I run  `tmt run plan --name plans/testcloud$` manually with that change, it pass.

here is the content of plans/testcloud.fmf

```
provision:
    how: virtual
    image: fedora-42

discover:
  how: shell
  # URL pointing to a public DistGit repository (required for fetching sources)
  url: https://src.fedoraproject.org/rpms/tmt 
  # Enable the core dist-git functionality
  dist-git-source: true 
  tests:
    - name: /print/pyproject/content
      # The test verifies the successful location and extraction of a file
      # (pyproject.toml) expected inside the top-level source directory.
      test: cat $TMT_SOURCE_DIR/tmt-*/pyproject.toml 
```

I'm okay with both email.message and werkzeug.http, gonna to approve this mr

**Review by skycastlelily — APPROVED** [review]

File: tmt/utils/url.py (hunk 3 above)
Comment (LecrisUT) [reply tmt/utils/url.py:21]: Good point about unit-testing. I can pick up some of them from there. I'll try to add it to this PR.

RE `werkzeug.http`, it looks quite promising. But it seems to bring in server functionality which might have us follow a few CVEs. Good find, can you open a follow-up issue to see which method to use.

Also, you might have more experience with this. Are there cases where the `Content-Disposition` only gives `filename*` without a `filename` backup?

Comment (skycastlelily) [reply tmt/utils/url.py:21]: >  Good find, can you open a follow-up issue to see which method to use.

sure, https://github.com/teemtee/tmt/issues/4255 please feel free to modify the content.

> Also, you might have more experience with this. 

Believe me, np , it's just that I feel bad about put the misleading comment, so I spend more time to find possible solution. 

> Are there cases where the Content-Disposition only gives filename* without a filename backup?

here is the reply from Gemini,but,yeah, I can't tell whether it's hallucination or not :sweat_smile:

Yes, there are cases where the Content-Disposition header may only include the filename* parameter without the legacy filename fallback.

While the general recommendation (especially for maximum compatibility with very old or non-standard clients) is to provide both, it is not mandatory under the current standard to include the filename parameter if filename* is present.

Comment (LecrisUT) [reply tmt/utils/url.py:21]: > > Also, you might have more experience with this.
> 
> Believe me, np , it's just that I feel bad about put the misleading comment, so I spend more time to find possible solution.

I mean more in that you probably interacted with websites that could have used this format more, since this is used for non-ascii filenames, Chinese characters being one of them. If randomly you can remember to try a manual curl command to check the headers occasionally, it would be a useful real-life datapoint.

File: tmt/steps/discover/fmf.py [hunk 4]
```diff
@@ -584,19 +585,11 @@ def go(self, *, logger: Optional[tmt.log.Logger] = None) -> None:
...
-                # TODO: Generalize this file download
-                if "Content-Disposition" in response.headers:
-                    archive_name = re.findall(
-                        "filename=(.+)", response.headers["Content-Disposition"]
-                    )[0]
-                else:
-                    archive_name = response.url.split("/")[-1]
-                archive_path = self.workdir / archive_name
-                archive_path.write_bytes(response.content)
+                assert self.workdir  # Narrow type
```
Comment (happz) [inline tmt/steps/discover/fmf.py]: A dedicated attribute should be used, most likely `self.phase_workdir`, to avoid `assert` as a type checker hint.

**Review by happz — APPROVED** [review]
